### PR TITLE
[Core] Account for Entity identifiers that are not simple strings

### PR DESCRIPTION
**Pull request by melodyogonna**
### **User description**
# Description

What - Entity identifiers can either be a string or a dict of search query, this was not being accounted for.

Why - Integrations were failing during reconciliation.

How - Simply using `str()` would work but order is not guaranteed, which is an issue for correctness. Instead I dump to json using sorted keys.

## Type of change

Please leave one option from the following and delete the rest:

- [ x ] Bug fix (non-breaking change which fixes an issue)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Core testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Resync finishes successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Scheduled resync able to abort existing resync and start a new one
- [ ] Tested with at least 2 integrations from scratch
- [ ] Tested with Kafka and Polling event listeners
- [ ] Tested deletion of entities that don't pass the selector

### Integration testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Completed a full resync from a freshly installed integration and it completed successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [ ] Handled rate limiting
- [ ] Handled pagination
- [ ] Implemented the code in async
- [ ] Support Multi account

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

___

### **PR Type**
Bug fix

___

### **Description**
- Fix entity identifier handling for dictionary types in reconciliation

- Convert dictionary identifiers to consistent JSON representation

- Improve type checking in EntityMapping class

- Add comprehensive test coverage for identifier scenarios

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Entity with dict identifier"] --> B["get_port_diff function"]
  B --> C["JSON dumps with sorted keys"]
  C --> D["Consistent string key"]
  D --> E["Successful reconciliation"]
  F["EntityMapping type check"] --> G["Use IngestSearchQuery type"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>models.py</strong><dd><code>Refine EntityMapping identifier type checking</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

port_ocean/core/handlers/port_app_config/models.py

<ul><li>Replace generic <code>dict</code> type check with specific <code>IngestSearchQuery</code> type<br> <li> Improve type specificity for search identifier detection</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2196/files#diff-80124dd7f8985d4a5f39ab391420f0784c19b3c424b54dbd80535c909aca329d">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>utils.py</strong><dd><code>Fix dictionary identifier handling in diff calculation</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

port_ocean/core/utils/utils.py

<ul><li>Add handling for dictionary identifiers in <code>get_port_diff</code> function<br> <li> Convert dict identifiers to JSON strings with sorted keys<br> <li> Ensure consistent key generation for entity comparison<br> <li> Import <code>json</code> module for serialization</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2196/files#diff-187b8268f2779c6fc01221ccc64a1bffcf8b5f5c96c5e951b4c8777fcd425d15">+18/-2</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_get_port_diff.py</strong><dd><code>Add comprehensive test coverage for get_port_diff</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

port_ocean/tests/core/utils/test_get_port_diff.py

<ul><li>Create comprehensive test suite for <code>get_port_diff</code> function<br> <li> Add test cases for dictionary identifiers, created/deleted/modified <br>entities<br> <li> Include helper function for test entity creation<br> <li> Cover edge cases and mixed change scenarios</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2196/files#diff-08e6642512b5b97acad592ceb713098fe9e7e718d797c0374c43e7c20846c3d8">+164/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Update changelog for version 0.28.8</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

CHANGELOG.md

<ul><li>Add version 0.28.8 release notes<br> <li> Document bug fixes for entity identifier handling<br> <li> Note EntityMapping type check improvement</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2196/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pyproject.toml</strong><dd><code>Version bump to 0.28.8</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pyproject.toml

- Bump version from 0.28.7 to 0.28.8

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2196/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: port_ocean/core/handlers/port_app_config/models.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ class EntityMapping(BaseModel):
 
     @property
     def is_using_search_identifier(self) -> bool:
-        return isinstance(self.identifier, dict)
+        return isinstance(self.identifier, IngestSearchQuery)
```
Comment (matan84) [inline port_ocean/core/handlers/port_app_config/models.py]:
```suggestion
        return isinstance(self.identifier, dict) or isinstance(self.identifier, IngestSearchQuery)
```

**Review by matan84 — APPROVED** [review]
LGTM, left 1 suggestion
